spacy/pipeline/_parser_internals/_state.pxd

```diff
@@ -41,25 +40,32 @@ cdef cppclass StateC:
     int offset
     int _b_i
 
-    __init__(const TokenC* sent, int length) nogil:
+    __init__(const TokenC* sent, int length) nogil except +:
+        cdef int init_capacity
+
+        this._heads.resize(length)
+        this._unshiftable.resize(length)
+
+        # Reserve memory ahead of time to minimize allocations during parsing.
+        # The initial capacity set here ideally reflects the expected average-case/majority usage.
+        init_capacity = 20
```

Comment:
Maybe this is too much of a micro-optimization, but since most non-Windows impls of `vector` use doubling, using a power of two will keep the capacities of power of two. This might be beneficial, since a lot of memory management use sizes that are powers of two (pages, different pools, etc).